From 2d5a13aa2b2516c588555dc9df0b833a97f4413c Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 16:14:07 -0400
Subject: [ethernet] be2net: re-enable vlan filtering mode asap

Message-id: <1410192881-30702-45-git-send-email-ivecera@redhat.com>
Patchwork-id: 92428
O-Subject: [RHEL7.1 PATCH 44/78] be2net: re-enable vlan filtering mode asap
Bugzilla: 1088575
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>

BZ: #1088575

Upstream commit(s):
commit 9d4dfe4ae370747bcab66531482013d7ff6857f1
Author: Kalesh AP <kalesh.purayil@emulex.com>
Date:   Mon Jun 30 13:01:33 2014 +0530

    be2net: re-enable vlan filtering mode asap

    While adding vlans, when the HW limit of vlan filters is reached, the
    driver enables vlan promiscuous mode.
    Similarily, while removing vlans, the driver must re-enable HW filtering
    as soon as the number of vlan filters is within the HW limit.

    Signed-off-by: Kalesh AP <kalesh.purayil@emulex.com>
    Signed-off-by: Sathya Perla <sathya.perla@emulex.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/emulex/benet/be_main.c b/drivers/net/ethernet/emulex/benet/be_main.c
index f7efe28..89f7dd6 100644
--- a/drivers/net/ethernet/emulex/benet/be_main.c
+++ b/drivers/net/ethernet/emulex/benet/be_main.c
@@ -1172,20 +1172,15 @@ static int be_vlan_add_vid(struct net_device *netdev, __be16 proto, u16 vid)
 static int be_vlan_rem_vid(struct net_device *netdev, __be16 proto, u16 vid)
 {
  struct be_adapter *adapter = netdev_priv(netdev);
- int status = 0;
 
  /* Packets with VID 0 are always received by Lancer by default */
  if (lancer_chip(adapter) && vid == 0)
-  goto ret;
+  return 0;
 
  clear_bit(vid, adapter->vids);
- status = be_vid_config(adapter);
- if (!status)
-  adapter->vlans_added--;
- else
-  set_bit(vid, adapter->vids);
-ret:
- return status;
+ adapter->vlans_added--;
+
+ return be_vid_config(adapter);
 }
 
 static void be_clear_promisc(struct be_adapter *adapter)
-- 
1.7.1